crates/noirc_evaluator/src/ssa/code_gen.rs

```diff
@@ -0,0 +1,1006 @@
+        for stmt_id in block.statements() {
+            self.evaluate_statement(env, stmt_id)?;
+        }
+
+        Ok(())
+    }
+
+    //Optimise, flatten and truncate IR and then generates ACIR representation from it
+    pub fn ir_to_acir(&mut self, evaluator: &mut Evaluator) -> Result<(), RuntimeError> {
+        //let mut number = String::new();
+
+        //SSA
+        dbg!("SSA:");
+        self.print();
+        //Optimisation
+        block::compute_dom(self);
+        dbg!("CSE:");
+        optim::cse(self);
+        self.print();
+        //Unrolling
+        dbg!("unrolling:");
+        flatten::unroll_tree(self);
+        optim::cse(self);
+        self.print();
+        //Truncation
+        integer::overflow_strategy(self);
+        self.print();
+        //println!("Press enter to continue");
+        //io::stdin().read_line(&mut number);
+        //ACIR
+        self.acir(evaluator);
+        dbg!("DONE");
+        Ok(())
+    }
+
+    fn evaluate_identifier(&mut self, env: &mut Environment, ident_id: &IdentId) -> arena::Index {
+        let ident_name = self.context.unwrap().def_interner.ident_name(ident_id);
+        let ident_def = self.context.unwrap().def_interner.ident_def(ident_id);
+        let var = self.find_variable(&ident_def); //TODO by name or by id?
+        let v_id;
+        if var.is_some() {
+            v_id = ssa_form::get_current_value(self, var.unwrap().id);
+            return v_id;
+        }
+        let obj = env.get(&ident_name);
+        let obj_type = node::ObjectType::get_type_from_object(&obj);
+
+        //new variable - should be in a let statement? The let statement should set the type
+        let obj = node::Variable {
+            id: self.id0,
+            name: ident_name.clone(),
+            obj_type,
+            root: None,
+            def: ident_def,
+            witness: node::get_witness_from_object(&obj),
+            parent_block: self.current_block,
+        };
+
+        v_id = self.add_variable(obj, None);
+        self.get_block_mut(self.current_block)
+            .unwrap()
+            .update_variable(v_id, v_id);
+        v_id
+    }
+
+    //Cast lhs into type rtype. a cast b means (a) b
+    fn new_cast_expression(&mut self, lhs: arena::Index, rtype: node::ObjectType) -> arena::Index {
+        //generate instruction 'a cast a', with result type rtype
+        let i = node::Instruction::new(
+            node::Operation::cast,
+            lhs,
+            lhs,
+            rtype,
+            Some(self.current_block),
+        );
+        self.add_object(node::NodeObj::Instr(i))
+    }
+
+    fn evaluate_infix_expression(
+        &mut self,
+        lhs: arena::Index,
+        rhs: arena::Index,
+        op: HirBinaryOp,
+    ) -> Result<arena::Index, RuntimeError> {
+        let ltype = self.get_object_type(lhs);
+
+        let optype = ltype;
+        //TODO if type differs try to cast them: check how rust handle this
```

Comment:
I may be reading this wrong, but rust doesn't have implicit casts between integer types. I think there are some optimization details concerning the Deref trait and conversion from function items to function types but none that seem relevant here.

Reply:
You are correct, we want to follow rust behavior so if there is no implicit cast then would the frontend report the error so I can assume the types are correct  here?

Reply:
I think in Noir's case, Kev is considering some implicit casting to a Field types or to a wider integer type with the same sign. I think I'd probably prefer it if the typechecker inserted these widening conversions itself into the Hir when it finds these cases. That way the backend wouldn't have to be concerned with these rules at all.
I'll double check with Kev on the expected behavior here and submit an issue for tracking.

Reply:
Sounds good to me, you can resolve this when you have checked with Kev.